### ISSUE TITLE
Difficulty Levels : Easy, Medium, Hard

### ISSUE THREAD

**Issue by p1839251** [performous/performous#570]
Hello,
It would be great if you could choose from different difficulty levels.
There's no need to modify the pitch detection algorithm but the displayed wave could be narrower (hard mode) or wider (easy mode), so you would have a more or less margin of error (as the wave covers more or less of the actual notes)
You can get points easier without singing in perfect pitch.

_(Some of my friends prefer Ultrastar over Performous because it has an option to choose a difficulty level and they have more success using an easier mode. I personally like if it's difficult so I'd use the hard mode which can be the current one or even a narrower wave... but this is totally up to you.)_

**Comment by Baklap4** [comment]
Is the song either tagged 'hard', 'medium', 'easy'? (in other words can you provide us an example?)

**Comment by Lord-Kamina** [comment]
> Is the song either tagged 'hard', 'medium', 'easy'? (in other words can you provide us an example?)

No; basically what he's asking for is the option to set a tolerance on the scoring system; for example, if you have to sing a B, on medium (Or easy) maybe you'd still get points if you did a C or an A#.

**Comment by p1839251** [comment]
> No; basically what he's asking for is the option to set a tolerance on the scoring system; for example, if you have to sing a B, on medium (Or easy) maybe you'd still get points if you did a C or an A#.

Yes, that's exactly what I meant.
I made an exaggerated visual representation by Gimp:
<img alt="preformous" src="https://i.imgur.com/PfxOv04.jpg" width=50%>
The thicker the wave is, the more notes can be covered
and you would still get some points even if you sing C when B should be sung

**Comment by Baklap4** [comment]
Ahh yes! that's very much possible to do i think 👍 
Will this be a setting which can be switched on during runtime (of the song) or beforehand (song selection) or just once for every song (settings menu)?

**Comment by p1839251** [comment]
So I've been thinking about this too. Quite a dilemma.
I wouldn't hide this feature in the settings menu but I wouldn't force people to choose a difficulty level before each song either. (It may irritate some users.) There would be a default difficulty setting (the current thickness), which could be visible in the song selection menu and could be changed if the player wants to. It would be saved once it's been changed so it wouldn't have to be changed every time. Yet another magic with Gimp:
<img alt="Gimp Image" src="https://i.imgur.com/m13Em9u.jpeg" height=50%>
What I think is important is to be able to choose different difficulty levels for individual players. (Each player should have a difficulty level, not each song*!) So either pitch wave can be thicker than the other when they compete against each other. However, this can be problematic with Performous as you don't have to specify the number of players.

*It's less likely that someone would jump between easy and hard levels than one of the players isn't as good as the others
Let me know what you think

**Comment by Tronic** [comment]
Yes, add a config option for easy mode (similar to karaoke mode). Just change also the max points to 1000 instead of 10000 so that the scores stay comparable and don't let your friends know that the maximum of hard mode is ten times more 🤣

Some background: request for difficulty modes is a long-time FAQ, asked many times here before. It hasn't been implemented because the feature doesn't make much sense when it only affects grading and makes *correct* singing actually more difficult (the note would be highlighted even when you are several notes off, so you don't see if you are on pitch or not).

**Comment by p1839251** [comment]
I both agree and disagree.
I personally love Performous. I think it's the best karaoke* game. I wouldn't change one thing.
However, it's not that popular as it should be! And I'm here to change this situation.

So, who's your target audience? Professional singers or programmers?!
My friends fall in neither of these groups. They can't sing in perfect pitch (none of them really want to) and won't modify config files.
They have their own lives, professions and all they want is a white lie about how good singers they are.
If it's a requested feature then probably many people struggling with pitches and I assume not every one of them wants to be the next 'American Idol'.
My experience is that they tend to use Ultrastar where you can lower the difficulty levels and can easier be a hit singer.
Especially if it's a party game... after a few alcoholic beverages, pitches are getting less and less important.

To sum up, I love Performous as it is. Even if you don't want 3 different difficulty levels, there could be a normal (current one) and an easier one, where you can get some points even if you don't hit the notes right. - I know pitch waves can be turned off (karaoke mode) but then you don't get scores either. (I still think wider waves is a good idea). If you really want to make a difference between the scores, you could and a trailing E (easy) at the end of the score if they 'cheated'. Eg. 5500 E
And hopefully we can win back some users!

_*I know it's not only a karaoke game_

**Comment by Baklap4** [comment]
Soo let's see to sum it all up:
* There will be 3 difficulty settings ranging from _easy_ to _medium_ to _hard_
* The difficulty settings can be set on a player basis, but first suffice with just a setting per song.
* The singing wave differs in height according to the difficulty.
* The scores for easy will top to 1000 instead of 10.000 to remain 'comparable'
** I don't think this is fair and meant to be `/sarcasm` correct me if i'm wrong @Tronic ?

As of current performous we have two things: Roughly accurate notes (indicated with the singing wave), Pitchcorrect notes (indicated with stars).

>So, who's your target audience? Professional singers or programmers?!

I don't think we have specified a target audience. We just program it for fun (incl new features) and sing along.

**Comment by Tronic** [comment]
A bit off topic but to make the game *popular*, it needs better graphics and built-in song downloader or other means of distribution. Graphics could be done in HDR (bloom effects on SDR displays), with juicier particle and popup effects, and entirely new freestyle note graphics. Starting the game for the first time with no songs, and having only a few obscure free songs available for download will drive away most first time users without them never bothering to look up what is available (and it doesn't help too much that new pop songs are not really available in any acceptable quality anyway - community notation tends not to be any good).

**Comment by Baklap4** [comment]
Hmm that's indeed a bit off-topic and some things are also mentioned in several issues
>  built-in song downloader or other means of distribution
#4 however this comes with legal issues if not done correctly (including some licensed songs, downloading licensed songs in background without user interference). Distributing songs (at least providing the interface to make it happen, utorrent like) is plausible and already has been worked on

**Comment by p1839251** [comment]
> * There will be 3 difficulty settings ranging from _easy_ to _medium_ to _hard_
> * The difficulty settings can be set on a player basis, but first suffice with just a setting per song. **
> * The singing wave differs in height according to the difficulty.
> * The scores for easy will top to 1000 instead of 10.000 to remain 'comparable'

It sounds like a perfect roadmap. But it's up to you.
I'm not a programmer, so I have no idea how difficult it is to implement. It's probably much easier to create only 2 different difficulty levels (easy and normal) and you could enable/disable it in the settings menu. It would already be much help for the less  artistically inclined people. (I've come across several ultrastar txt files that were not even in the same key as the song!)

**Do you think people would use different settings on different songs? I don't think one would say, this song is easy I'll use hard mode here but that one is way more difficult let's set it to easy. I think there are more and less talented people and they'd stick to one of the options.

_(Maybe my taste is not so good but I'm quite satisfied with its graphics.)_

**Comment by Baklap4** [comment]
@p1839251 with #587 merged in current master we can now define difficulty levels.
* This can be done within the settings menu
* The setting is used globally
* There are 3 difficulties to choose from: Normal, Hard, Perfect
* Difficulty is also saved into the highscores
* Singing waves are thinner and thinner 
* Singing bars are thinner and thinner
* Max points still remain 10.000

For more technical info please see #587

If this resolves your issue please close the issue. Else the issue will be autoclosed after 7 days.

**Comment by ooshlablu** [comment]
So FWIW from my experience in playing Rock Band in decent size groups, people will use different difficulty settings on a per-song basis, and will do the same for instruments as well. I'm not sure if this is a bug, but when I play any song, I have to hit enter on a keyboard to select that I'm singing vocals anyway, so I don't think it would be a big deal to have a difficulty prompt there. The same behavior happens when starting a song and having instruments join anyway.

Again, from experience with Rock Band, I actually like being prompted up-front for difficulty, especially with how it is for the instruments already. You have no idea how many times I've had to abandon a song in RB because someone didn't set their difficulty correctly lol

**Comment by Baklap4** [comment]
>  I'm not sure if this is a bug, but when I play any song in Performous, I have to hit enter on a keyboard to select that I'm singing vocals anyway

@ooshlablu That shouldn't be the case. Unless you have a duet song or a combi of karaoke/band/ddr. So please file a different issue for that

> Again, from experience with Rock Band, I actually like being prompted up-front for difficulty, especially with how it is for the instruments already. You have no idea how many times I've had to abandon a song in RB because someone didn't set their difficulty correctly lol

@ooshlablu I think this would be nice to have yes. If that's the only thing we should close this issue, create a new issue and reference this issue with possibly the pr to make it nice and clean.

**Comment by Baklap4** [comment]
Closing this in favor of #654
